src/MergeContext.js

```diff
@@ -859,107 +962,175 @@ class MergeFinalizer extends MergeContext {
 
-        assert(compareStatus === "ahead");
-
         const stagingResult = await this._processStagingStatuses();
         if (!stagingResult.succeeded())
             return stagingResult;
 
         if (await this._stagingOnly()) {
-            await this._labelPassedStagingChecks();
+            this._labelPassedStagingChecks();
             return StepResult.Suspend();
         }
         return StepResult.Succeed();
     }
 
     async _mergeToBase() {
         assert(this._tagSha);
+        assert(this._compareStatus === "ahead");
         this._log("merging to base...");
         try {
             await GH.updateReference(this._prBaseBranchPath(), this._tagSha, false);
+            this._prState = PrState.PostStaged();
             return StepResult.Succeed();
         } catch (e) {
             if (e.name === 'ErrorContext' && e.unprocessable()) {
                 if (await this._tagDiverged()) {
                     Log.LogException(e, this._toString() + " fast-forwarding failed");
-                    return await this._cleanupMergeFailed(true);
+                    return await this._cleanupMergeFailed();
                 }
             }
             throw e;
         }
     }
 
+    async _acquireUserProperties() {
+        const emails = await GH.getUserEmails();
+        for (let e of emails) {
+            if (e.primary) {
+                Config.githubUserEmail(e.email);
+                break;
+            }
+        }
+        assert(Config.githubUserEmail());
+
+        const user = await GH.getUser(Config.githubUserLogin());
+        Config.githubUserName(user.name);
+        assert(Config.githubUserName());
+    }
+
+    async _createStaged() {
+        const baseSha = await GH.getReference(this._prBaseBranchPath());
+        const mergeSha = await GH.getReference("pull/" + this._prNumber() + "/merge");
+        const mergeCommit = await GH.getCommit(mergeSha);
+        if (!Config.githubUserName())
+            await this._acquireUserProperties();
+        let now = new Date();
+        const committer = {name: Config.githubUserName(), email: Config.githubUserEmail(), date: now.toISOString()};
+        const tempCommitSha = await GH.createCommit(mergeCommit.tree.sha, this._prMessage(), [baseSha], mergeCommit.author, committer);
+        this._tagSha = await GH.createReference(tempCommitSha, "refs/" + this._stagingTag());
+        this._compareStatus = "ahead";
+        await this._setApprovalStatus(this._tagSha);
+        await GH.updateReference(Config.stagingBranchPath(), this._tagSha, true);
+    }
+
+    // returns filled StepResult object
+    async _stage() {
+        this._role = "stager";
+
+        this._unlabelPreconditionsChecking();
+
+        const conditions = await this._checkPreconditions();
+
+        if (!conditions.succeeded()) {
+            if (conditions.delayed())
+                return conditions;
+            // for failed() or suspended()
+            return StepResult.Fail();
```

Comment:
> we cannot, since the 'suspended' is used by the higher level code to take the decision whether a PR is being staged

Please fix that (i.e., fix process() caller to use PullRequest::staged() to determine whether the PR is staged) and then come back to this change request. I see no reason to assign a special meaning to the "suspended" result of a process() call. Suspended result should mean that the requested operation could not complete immediately and is waiting for some external non-time event. We should not piggyback "or the PR got staged" on top of that.